Fix run nohw

dpino opened the pull request:
The run_nohw, the app that lets run an lwAFTR over network interfaces, was broken. This PR fixes the app and adds other improvements. Check out the commit messages for a description of what each commit does.

How to use run_nohw app:

First thing is to add two veth pairs. A veth pair is like a patch cable. What is copied in one end arrives the other end:

```bash
$ sudo ip link add veth0 type veth peer name veth1
$ sudo ip link add veth2 type veth peer name veth3
```

Set the interfaces up:

```bash
$ sudo ip set veth0 up
$ sudo ip set veth1 up
$ sudo ip set veth2 up
$ sudo ip set veth3 up
```

Run the lwAFTR with run_nohw:

```bash
$ sudo ./snabb lwaftr run_nohw -v --conf lwaftr-migrated.conf --inet-if veth1 --b4-if veth3
```

Send traffic via the other veth pairs:

```bash
$ sudo tcpreplay -i veth0 from-inet-0550.pcap
```

When traffic hit the lwAFTR interface, run_nohw will report accordingly:

```
Time (s),IPv4 TX MPPS,IPv4 TX Gbps,IPv4 RX MPPS,IPv4 RX Gbps,IPv6 TX MPPS,IPv6 TX Gbps,IPv6 RX MPPS,IPv6 RX Gbps
0.999632,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
1.999666,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
2.999727,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
3.999644,0.000770,0.003388,0.000000,0.000000,0.000000,0.000000,0.000770,0.003635
```

Tap interfaces are much slower than NICs, so the result above is normal.  The goal of this app is to provide a means to test out the lwAFTR, specially for those users lacking an Intel82599 card.

dpino committed:
Fix run_nohw

* Add lwAFTR's hairpinning interface.
* Move engine.configure before CSVStats setup.
* Set correct apps names is CSVStats.

dpino committed:
Clean up run_nohw

* Replace semicolons for commas
* Remove verbosity level 2 in run_nohw. Involved apps in this graph
(RawSocket and lwAFTR) do not overwrite their report() method, so nothing
is actually reported.

dpino committed:
Add option 'duration' in run_nohw

* Link statistics are only shown up if the program finishes, thus a
duration is needed.
* Refactor options in run_nohw.

dpino committed:
Add run_nohw to the list of subprograms

Clean up list and add run_nohw.

dpino committed:
Update run_nohw README

dpino committed:
Rework verbosity option

dpino committed:
Fix pid in CSVStats

kbara commented:
I think this needs a regression test. Thoughts, @teknico ?

teknico commented:
Yes, it does. We don't even have a subcommand test to be invoked by CI, for some reason. If we had one, I guess we'd notice the brokenness earlier. :-)

dpino commented:
SGTM, I will add an UI test for run_nohw too.

teknico commented on src/program/lwaftr/run_nohw/README:
```diff
@@ -4,14 +4,15 @@ Usage: run-nohw --help
        --conf, -c PATH      Path to the lwAFTR configuration file.
```
(Github won't let me add this comment two lines above, where it should go.)

The main `README` and all example mention `run_nohw` with an underscore, rather than with a dash as shown here.

dpino replied:
Dash is also valid as a program or a subprogram name. Some context, Lua's filenames or modules cannot contain dashes, so you will see for instance that all the folders in programs/, whenever there's a compound name, use _ . Snabb's engine contains some logic to replace a module or submodule name that contains '-' to _. And that's why it works too.

teknico replied:
Uhm. If that's how upstream does it, ok, but I don't especially like it. At least for our code, I'd rather stick to one way and use and document it all over the place.

dpino replied:
OK, I will replace instances of "run_nohw" for "run-nohw" in the docs.

teknico commented on src/program/lwaftr/run_nohw/README:
```diff
@@ -4,14 +4,15 @@ Usage: run-nohw --help
+                            written.  A simple filename or relative pathname
+                            will be based on the current directory.
        --help, -h           Show this help message.
 
```
We need a description of the subcommand purpose and behavior here.

teknico commented on src/program/lwaftr/run_nohw/run_nohw.lua:
```diff
@@ -17,45 +17,49 @@ local function check(flag, fmt, ...)
          check(file_exists(arg), "no such file '%s'", arg)
          conf_file = arg
-      end;
+      end,
```
Not sure why the semicolons worked before. Or they did not? :-)

dpino replied:
They worked. Lua allows to separate table items with semicolons but it's not common. In most of the codebase, commas are used instead.

I think it was @aperezdc who wrote this code. Maybe he can explain better what was the reason for using semicolons :)

aperezdc replied:
There was no particular reason for semicolons, other than usually for array-shaped tables I tend to use commas as separators, and for dictionary-shaped tables I tend to use semicolons — that's all. I am aware that the coding style for Snabb is using always commas, so this probably has slipped in unnoticed.

teknico commented on src/program/lwaftr/run_nohw/run_nohw.lua:
```diff
@@ -73,24 +77,20 @@ function run(parameters)
+   if opts.duration then
+      engine.main({duration = opts.duration, report = { showlinks = true }})
+   else
+      engine.main()
```
Now links are not shown anymore, if the duration option is not used. Not sure how the two things are related.

dpino replied:
Links report were not shown before, because the program run indefinitely by default. The point is that unless a duration is set in `engine.main()` the program never finishes. If the program never finishes, reports are never shown (despite they were set).

The bottom line is to show a links report or an apps report, it's necessary to set a duration. If duration is not set just simply run the program indefinitely without setting any report option.

teknico replied:
Ah, ok, in that case it makes sense.

teknico reviewed (commented):
This needs a test, as mentioned above.

dpino committed:
Add unit test for run_nohw

dpino committed:
Update documentation

teknico commented on src/program/lwaftr/tests/subcommands/run_nohw_test.py:
```diff
@@ -0,0 +1,77 @@
+    program = [
+        str(SNABB_CMD), 'lwaftr', 'run_nohw',
+    ]
+    cmd_args = {
```
I like this separation between arguments, that have to be in the right order, and options, that have no such constraint, and can be better represented by a dictionary. I would rename `program` to `cmd_args` and `cmd_args` to `cmd_options`, though,

teknico commented on src/program/lwaftr/tests/subcommands/run_nohw_test.py:
```diff
@@ -0,0 +1,77 @@
+        cls.create_veth_pair()
+
+    @classmethod
+    def create_veth_pair(cls):
```
Since `create_veth_pair` is only called by `setUpClass`, it probably doesn't need to be a separate classmethod. What it does is explained by comments anyway.

teknico commented on src/program/lwaftr/tests/subcommands/run_nohw_test.py:
```diff
@@ -0,0 +1,77 @@
+        cls.remove_veths()
+
+    @classmethod
+    def remove_veths(cls):
```
As above, since `remove_veths` is only called by `tearDownClass`, it also probably doesn't need to be a separate classmethod.

teknico commented on src/program/lwaftr/tests/subcommands/run_nohw_test.py:
```diff
@@ -0,0 +1,77 @@
+        check_call(('ip', 'link', 'set', veth0, 'up'))
+        check_call(('ip', 'link', 'set', veth1, 'up'))
+
+        # Add interface names to class.
```
I don't especially like so many blank lines inside a  function or method. PEP8 does [insist](https://www.python.org/dev/peps/pep-0008/#blank-lines) on "sparingly". :-)

teknico commented on src/program/lwaftr/tests/subcommands/run_nohw_test.py:
```diff
@@ -0,0 +1,77 @@
+        return 'veth%s' % randint(10000, 999999)
+
+    def test_run_nohw(self):
+        self.execute_run_test(self.cmd_args)
```
In other tests I created a separate `execute_run_test` method when there was common code to a number of tests. As long as there's just one, it probably doesn't need to be separated.

teknico commented on src/program/lwaftr/tests/subcommands/run_nohw_test.py:
```diff
@@ -0,0 +1,77 @@
+        result = self.program
+        for item in self.cmd_args.items():
+            for each in item:
+                result.append(each)
```
These three lines may be written more shortly as:
```
        for key, value in self.cmd_args.items():
            result.extend((key, value))
```

teknico commented on src/program/lwaftr/tests/subcommands/run_nohw_test.py:
```diff
@@ -0,0 +1,77 @@
+    @classmethod
+    def remove_veths(cls):
+        for i in range(0, len(cls.veths), 2):
+            check_call(('ip', 'link', 'delete', cls.veths[i]))
```
These two lines can be simplified as:
```
        for veth in cls.veths:
            check_call(('ip', 'link', 'delete', veth))
```

dpino replied:
To delete a veth pair is just necessary to delete one of the ends, otherwise an error is returned when trying to remove both. This was the reason why the loop iterated even numbers. In any case, as there's only two elements in the array as the test is now I removed the loop completely and just remove veth[0].

teknico reviewed (changes requested):
Thank you for adding the test, and sorry for all the nit-picking. :-)

I wonder if we should also add support for the `--reconfigurable` option, and how to keep this in sync with the `run` subcommand in general.

dpino committed:
Rework unit test

dpino commented:
Thanks for the tips, they were very valuable. I think I have addressed all the comments. As for supporting the reconfigurable option, although interesting (specially for running "snabb config" commands), I would leave it for now. 

As for keeping it in sync with the "run" command, some time ago (actually months) @aperezdc pushed a PR that added support for tap interfaces in "lwaftr run", so this subcommand was no longer needed. The lwAFTR grew in complexity and the PR got out of sync soon. Adrian's PR could be retaken in the future or write the code again  change again from scratch. On the other hand, I think having it as a separated command makes the script more manageable.

teknico commented:
Nice, thanks for the changes.
